lib/Sema/BuilderTransform.cpp

```diff
@@ -1226,13 +1206,14 @@ class BuilderClosureRewriter
       // Skip variable declarations; they're always part of a pattern
       // binding.
       if (isa<VarDecl>(decl)) {
+        TypeChecker::typeCheckDecl(decl);
         newElements.push_back(decl);
         continue;
       }
 
       // Handle pattern bindings.
       if (auto patternBinding = dyn_cast<PatternBindingDecl>(decl)) {
-        finishPatternBindingDecl(patternBinding);
+        TypeChecker::typeCheckDecl(decl);
```

Comment:
I think we should keep `finishPatternBindingDecl` and call `typeCheckDecl` afterward to prevent re-typechecking initializers (because we did that already while working on the body).

Reply:
And, if I remember correctly, you can inline `finishPatternBindingDecl` into getting a solution application target for the whole `patternBinding` instead of individual components and `rewriteTarget` on that would do the right thing for each of its entries.

Reply:
typeCheckDecl will not attempt to re-typecheck anything that's already been validated before. Pattern bindings have the "initializer checked" bit that should be set during solution application.

Reply:
> Pattern bindings have the "initializer checked" bit that should be set during solution application.

That's exactly what is happening in `finishPatternBindingDecl` via `rewriteTarget` that you have removed.

Reply:
That's why I'm saying that we should use `patternBinding` itself as a solution target for rewrite so it could set the bits correctly in here - https://github.com/apple/swift/blob/main/lib/Sema/CSApply.cpp#L8626-L8651

Reply:
That same code path should be invoked by the decl checker. 

```
TypeCheckDecl(PBD) -> visitPatternBinding -> typeCheckBinding -> typeCheckExpression -> rewriteTarget
```

Reply:
But we don't want to do that though - pattern has been type-check as part of the result builder body together with enclosing context here, and it's already a part of solution, so it could be applied and flags set so all `typeCheckDecl` needs to do is to validate internal state.

Reply:
Alright, I see what you mean. I'll put `finishPatternBindingDecl` back and call typeCheckDecl afterwards to come in and do the rest.